subsys/net/lib/http/http_app_server.c

```diff
@@ -0,0 +1,1085 @@
+
+		net_addr_ntop(addr->sa_family,
+			      &net_sin6(addr)->sin6_addr,
+			      ipaddr, sizeof(ipaddr));
+		snprintk(buf, buflen, "[%s]:%u", ipaddr,
+			 ntohs(net_sin6(addr)->sin6_port));
+#endif
+	} else if (addr->sa_family == AF_INET) {
+#if defined(CONFIG_NET_IPV4)
+		char ipaddr[NET_IPV4_ADDR_LEN];
+
+		net_addr_ntop(addr->sa_family,
+			      &net_sin(addr)->sin_addr,
+			      ipaddr, sizeof(ipaddr));
+		snprintk(buf, buflen, "%s:%u", ipaddr,
+			 ntohs(net_sin(addr)->sin_port));
+#endif
+	} else {
+		snprintk(buf, buflen, "<invalid address family %d>",
+			 addr->sa_family);
+	}
+
+	return buf;
+}
+
+static struct net_context *get_server_ctx(struct net_app_ctx *ctx,
+					  const struct sockaddr *dst)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_NET_APP_SERVER_NUM_CONN; i++) {
+		struct net_context *tmp;
+		u16_t port, rport;
+
+		if (!ctx->server.net_ctxs[i]) {
+			continue;
+		}
+
+		tmp = ctx->server.net_ctxs[i];
+
+		if (IS_ENABLED(CONFIG_NET_IPV4) &&
+		    tmp->remote.sa_family == AF_INET &&
+		    dst->sa_family == AF_INET) {
+			struct in_addr *addr4 = &net_sin(dst)->sin_addr;
+			struct in_addr *remote4;
+
+			remote4 = &net_sin(&tmp->remote)->sin_addr;
+			rport = net_sin(&tmp->remote)->sin_port;
+			port = net_sin(dst)->sin_port;
+
+			if (net_ipv4_addr_cmp(addr4, remote4) &&
+			    port == rport) {
+				return tmp;
+			}
+
+		} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
+			   tmp->remote.sa_family == AF_INET6 &&
+			   dst->sa_family == AF_INET6) {
+			struct in6_addr *addr6 = &net_sin6(dst)->sin6_addr;
+			struct in6_addr *remote6;
+
+			remote6 = &net_sin6(&tmp->remote)->sin6_addr;
+			rport = net_sin6(&tmp->remote)->sin6_port;
+			port = net_sin6(dst)->sin6_port;
+
+			if (net_ipv6_addr_cmp(addr6, remote6) &&
+			    port == rport) {
+				return tmp;
+			}
+		}
+	}
+
+	return NULL;
+}
+#endif /* CONFIG_NET_DEBUG_HTTP */
+
+static inline void new_client(struct http_ctx *ctx,
+			      enum http_connection_type type,
+			      struct net_app_ctx *app_ctx)
+{
+#if defined(CONFIG_NET_DEBUG_HTTP) && (CONFIG_SYS_LOG_NET_LEVEL > 2)
+#if defined(CONFIG_NET_IPV6)
+#define PORT_LEN sizeof("[]:xxxxx")
+#define ADDR_LEN NET_IPV6_ADDR_LEN
+#elif defined(CONFIG_NET_IPV4)
+#define PORT_LEN sizeof(":xxxxx")
+#define ADDR_LEN NET_IPV4_ADDR_LEN
+#endif
+	char buf[ADDR_LEN + PORT_LEN];
+	struct net_context *net_ctx;
+	const char *type_str = "HTTP";
+
+	net_ctx = get_server_ctx(app_ctx, ctx->addr);
+	if (net_ctx) {
+		NET_INFO("[%p] %s connection from %s (%p)", ctx, type_str,
+			 sprint_ipaddr(buf, sizeof(buf), &net_ctx->remote),
+			 net_ctx);
+	} else {
+		NET_INFO("[%p] %s connection", ctx, type_str);
+	}
+#endif /* CONFIG_NET_DEBUG_HTTP */
+}
+
+static void url_connected(struct http_ctx *ctx,
+			  enum http_connection_type type)
+{
+	new_client(ctx, type, &ctx->app_ctx);
+
+	if (ctx->cb.connect) {
+		ctx->cb.connect(ctx, type, ctx->user_data);
+	}
+}
+
+struct http_root_url *http_server_add_url(struct http_server_urls *my,
```

Comment:
If the URLs array is sorted by URL length, it's possible to quickly perform a longest subprefix match without requiring fancier data structures like a trie.  This should simplify URL matching as well.

Reply:
This can be done by application which can fill the url array in desired order.